### ISSUE TITLE
Generated Functions inside macros

### ISSUE THREAD

**Issue by fcard** [JuliaLang/julia#18747]
```julia
macro m(x)
  @generated f() = x
  :($(f)())
end
```
```julia
julia> @m 1
ERROR: type DataType has no field x
 in f(...) at ./REPL[1]:2
```
As performed in `0.5.0` and `0.6.0-dev.826`.
:?
```julia
julia> @code_lowered macroexpand(:(@m 1)).args[1]()
CodeInfo(:(begin 
        nothing
        return (Core.getfield)(#self#,:x)
    end))
```

**Comment by simonster** [comment]
This issue seems to apply to inner generated functions generally:

```julia
julia> function f(x)
       @generated g() = x
       g
       end;

julia> f(1)()
ERROR: type DataType has no field x
 in g(...) at ./REPL[1]:2
```

**Comment by simonster** [comment]
This may be similar to #15602 and #18674, in that it is no longer permitted but doesn't have an appropriate error message.

**Comment by fcard** [comment]
Oh, that's nice to know. I just tried something like this and it worked:
```julia
function f(x)
  @generated function()
    x
  end
end
```
```julia
julia> f(1)()
1
```
So I assumed it was a macro thing. I will update the issue, then.

> This may be similar to #15602 and #18674, in that it is no longer permitted but doesn't have an appropriate error message.

Maybe? Also like #16430. Tbh, sometimes metaprogramming in julia feels like type parameters in java (damn you type erasure!), but I *guess* I will survive, hehe. I really need to take some time to understand the compiler better and see what the conceptual limitations are...

**Comment by vtjnash** [comment]
I'm not sure how easily this can be fixed, but the correct answer should have been `Int` in all of the above examples (since the function generator capture types, not values).

**Comment by fcard** [comment]
@vtjnash yeah, I later realized that the above gave a `1` because `@generated function() ... end` doesn't create a generated function at all, it just returns the function itself :P That's a separate issue I guess, is that reported anywhere? (I am guessing `@generated` just doesn't support anonyfuncs... yet??)

```julia
f = @generated function() end
g = @generated function h() end
```
```julia
julia> methods(f).ms[1].isstaged
false

julia> methods(g).ms[1].isstaged
true
```

**Comment by fcard** [comment]
@vtjnash hey wait, you sure? The "`@generated function() end` doesn't work" issue is true, but the "closure captures types" isn't consistent with other outside variables:
```julia
x = 1
@generated f() = x
```
```julia
julia> f()
1

julia> methods(f).ms[1].isstaged
true
```

**Comment by JeffBezanson** [comment]
Yes, I think it would be reasonable for generated functions to still capture variables by value. Arguably `@generated` only causes the generator's arguments to be bound to types, and nothing else.

**Comment by vtjnash** [comment]
True, this is perhaps one way we could allow inner generated functions to capture variables, by requiring that the generator also reference that variable? The implementation of such sounds quite annoying.

However, the following would need to hold in order for the cached code to have only depended on the type `#g#f{typeof(x)}`, which is why in the above example, the correct answer would be `Int`:

```julia
function g(x)
   @generated f()
       return :(x :: $x)
   end
   return f
end
```

**Comment by vtjnash** [comment]
In your code example,
```julia
x = 1
@generated f() = x
```
you are missing a `const` declaration before the `x` for this to be a valid usage of `generated` functions. This is not capturing `x`, but creating a new const reference to it.

**Comment by Keno** [comment]
This is now a lowering error:
```
julia> f(x) = @generated g() = x
ERROR: syntax: Global method definition around REPL[3]:1 needs to be placed at the top level, or use "eval".
Stacktrace:
 [1] top-level scope
   @ REPL[3]:1
```
